test/modules/fundingManager/bondingCurveFundingManager/ToposBondingCurveFundingManager.t.sol

```diff
@@ -127,10 +183,1571 @@ contract ToposBondingCurveFundingManagerTest is ModuleTest {
+        │   └── When: the function _redeemTokensFormulaWrapper() gets called
+        │       └── Then: it should return (capitalAvailable - MIN_RESERVE) instead of redeemAmount
+        └── Given: (capitalAvailable - redeemAmount) >= MIN_RESERVE
+            └── When: the function _redeemTokensFormulaWrapper() gets called
+                └── Then: it should return redeemAmount
+    */
+
+    function testInternalRedeemTokensFormulaWrapper_revertGivenCapitalAvailableIsZero(
+    ) public {
+        uint _depositAmount = 1;
+
+        // Transfer all capital that is in the bonding curve funding manager
+        vm.prank(address(bondingCurveFundingManager));
+        _token.transfer(seller, MIN_RESERVE);
+
+        // Execute Tx
+        vm.expectRevert(
+            IToposBondingCurveFundingManager
+                .ToposBondingCurveFundingManager__NoCapitalAvailable
+                .selector
+        );
+        bondingCurveFundingManager.call_redeemTokensFormulaWrapper(
+            _depositAmount
+        );
+    }
+
+    function testInternalRedeemTokensFormulaWrapper_worksGivenItReturnsCapitalAvailableMinusMinReserve(
+        uint _depositAmount
+    ) public {
+        // protect agains overflow
+        _depositAmount = bound(_depositAmount, 1, 1e36 - MIN_RESERVE);
+
+        // Calculate expected return value. Because capital available == MIN_RESERVE, any redeem amount
+        // triggers a change in return value as the balance may never dip below MIN_RESERVE, making the return
+        // value 0
+        uint expectedReturnValue =
+            bondingCurveFundingManager.call_getCapitalAvailable() - MIN_RESERVE;
+
+        // Get return value
+        uint functionReturnValue = bondingCurveFundingManager
+            .call_redeemTokensFormulaWrapper(_depositAmount);
+
+        // Assert equal
+        assertEq(functionReturnValue, expectedReturnValue);
+    }
+
+    function testInternalRedeemTokensFormulaWrapper_worksGivenItReturnsRedeemAmount(
+        uint _depositAmount
+    ) public {
+        // protect agains overflow
+        _depositAmount = bound(
+            _depositAmount,
+            1,
+            1e36 - bondingCurveFundingManager.call_getCapitalAvailable()
+        );
+        _mintCollateralTokenToAddressHelper(
+            address(bondingCurveFundingManager), _depositAmount
+        );
+
+        // Get expected return value
+        uint redeemAmount = IToposFormula(formula).tokenIn(
+            _depositAmount,
+            bondingCurveFundingManager.call_getCapitalAvailable(),
+            bondingCurveFundingManager.basePriceToCapitalRatio()
+        );
+        // Get return value
+        uint functionReturnValue = bondingCurveFundingManager
+            .call_redeemTokensFormulaWrapper(_depositAmount);
+
+        // Because of precision loss, the assert is done to be in range of 0.0000000001% of each other
+        assertApproxEqRel(functionReturnValue, redeemAmount, 0.0000000001e18);
```

Comment:
@marvinkruse I could not get both values equal, which seemed odd to me, as the precision loss should be the same if there is any. Would you know why? The diff is very minimal which I resolved in the testing like this

Reply:
Hm.. weird indeed, I expected it to be the same as well. DId you check what the difference is when it appears? Is it just 1 wei or anything bigger than that? Then at least it might be apparent that it's a clear precision loss from the division in for example the `basePriceToCapitalRatio`. But not sure why it would lead to different results, both is done in Solidity with the same values. Or is the order in which these calcs happen different?

I will check again in the morning after a coffee, maybe I see it then..  In any case, the solution you have here is good enough for me 😁

Reply:
So the difference was not a constant but changed with the fuzzing, and as I tried to define the allowed delta. So the value I now came up with works, but it's indeed weird because basically the same function with the same values get called

Reply:
Hm.. weird indeed. I don't think this should block the PR though, even if I'd like to understand this eventually. Let's go with the solution you have here and maybe either get the auditors to verify this or revisit this on our own down the road (before we deploy to mainnet).